hack/generator/pkg/codegen/pipeline_status_augment.go

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package codegen
+
+import (
+	"context"
+
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
+)
+
+func augmentSpecWithStatus() PipelineStage {
+	return MakePipelineStage(
+		"augmentSpecWithStatus",
+		"Merges information from Status into Spec",
+		func(ctx context.Context, types astmodel.Types) (astmodel.Types, error) {
+
+			newTypes := make(astmodel.Types)
+			newTypes.AddTypes(types)
+
+			// build the augmenter we will use:
+			augmenter := fuseAugmenters(
+				flattenAugmenter(newTypes), // note that this can alter newTypes as a side-effect, so it must be preloaded with all the types
+			)
+
+			for typeName, typeDef := range newTypes {
+				resource, ok := typeDef.Type().(*astmodel.ResourceType)
+				if !ok {
+					continue
+				}
+
+				// augment spec with any bits needed from status
+				newSpec, err := augmenter(resource.SpecType(), resource.StatusType())
+				if err != nil {
+					return nil, err
+				}
+
+				newTypes[typeName] = typeDef.WithType(resource.WithSpec(newSpec))
+			}
+
+			return newTypes, nil
+		})
+}
+
+// an augmenter adds information from the Swagger-derived type to
+// the main JSON schema-derived type, and returns the new type
+type augmenter func(main astmodel.Type, swagger astmodel.Type) (astmodel.Type, error)
+
+// fuseAugmenters merges multiple augments into one by applying each
+// augmenter in order to the result of the previous augmenter
+func fuseAugmenters(augments ...augmenter) augmenter {
+	return func(main astmodel.Type, swagger astmodel.Type) (astmodel.Type, error) {
+		var err error
+		for _, augment := range augments {
+			main, err = augment(main, swagger)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		return main, nil
+	}
+}
+
+// flattenAugmenter copies across the "flatten" property from Swagger
+func flattenAugmenter(allTypes astmodel.Types) augmenter {
```

Comment:
Wondering if there are other things in the future we want to merge in from the Swagger types, will we need to duplicate most of this main type vs Swagger type (main prop vs swagger prop) logic?

Maybe when we get a second one of these we should refactor this to have a pluggable last bit for the last method (line 114?) Or maybe you pull the first few anonymous functions out into named functions so that if we do end up with other impls they look like:

```
merger.Add(defaultType)
merger.Add(defaultTypeName)
merger.Add(lookupSwaggerType)
merger.Add(... // Your custom handler here)
```

Reply:
Agree that we will probably want to pull parts of this out to be shared. But I won't do it in this PR until we have a use-case where we know what we want.